darkman.1.scd
darkman(1)

# NAME

darkman - daemon for dark-mode and light-mode transitions on Unix-like desktops

# SYNOPSIS

*darkman* _run_++
*darkman* _set_ [_light_|_dark_]++
*darkman* _get_++
*darkman* _toggle_

# DESCRIPTION

*darkman* runs in the background and turns on dark mode at sundown, and turns it off
again at sunrise. darkman is not designed to be used interactively: it's designed to
be set up once, and run in the background.

It is also possible to trigger manual transitions and it is also possible to
disable automatic transitions entirely.

# COMMANDS

*run*
	Runs the darkman service. This command is intended to be executed by a
	service manager, init script or alike.

*set* <light|dark>
	Sets the current mode.

*get*
	Prints the current mode.

*toggle*
	Toggle the current mode.

# INTEGRATIONS

The open source desktop ecosystem is quite heterogeneous and making different
applications switch between dark/light different mechanism and techniques.

Darkman seeks to implement the more widely adopted standards, while leaving room
for users to hook in custom scripts for other applications.

## Custom executables

For any sort of custom integration, executables (including simple shell scripts)
can be placed in the following two directories:

- _$XDG_DATA_DIRS/dark-mode.d/_: Executed when switching to dark mode.
- _$XDG_DATA_DIRS/light-mode.d/_: Executed when switching to light mode.

These scripts or executables can perform any actions required, like telling
re-writing configuration files for a PDF reader, or controlling a notification
daemon to switch to another theme.

Scripts need to have an executable bit set, or will not be executed.

The variable `$XDG_DATA_DIRS` is defined in the xdg basedir specification, and
usually matches the following, amongst others:

- ~/.local/share/
- /usr/local/share/
- /usr/share/

Example scripts (and discussion on how to integrate different applications) are
available in the project repository:

	https://gitlab.com/WhyNotHugo/darkman

Packages may also drop-in their own scripts into any of these locations,
although application developers are encouraged to use the D-Bus API to
determine the current mode and listen for changes (see below for details).

## XDG Settings portal

Darkman implements the XDG desktop portal's dark mode standard. Applications
using this API should switch to dark/light mode based on darkman's current
preference. This standard was originally pushed by the GNOME and Elementary
teams, and is currently supported by KDE, Firefox and many other projects. You
should expect applications from those environment to support it, amongst others.

For more details on this protocol, see:

	https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Settings.html

As for *xdg-desktop-portal* version 1.17.0, portals MUST be configured with
per-user configuration *portals.conf(5)*. To force the usage of darkman for
dark/light mode setting, use something like the following:

```
[preferred]
org.freedesktop.impl.portal.Settings=darkman
```

When using a desktop-specific configuration (e.g.: *swaywm-portals.conf*),
please keep in mind that the environment variable *XDG_CURRENT_DESKTOP* must be
set for the *xdg-desktop-portal*.

The *xdg-desktop-portal* should start after *darkman* has started and is ready.
Use *--ready-fd* for readiness notification. This is likely not relevant on
systemd-based setups, where the service manager intermediates in taking the
named bus.

For a more in-depth explanation, see this article:

	https://whynothugo.nl/journal/2024/04/09/darkman-portal-configuration/

## D-Bus API

For custom integrations, darkman exposes a D-Bus API which allows querying and
controlling the current mode. The *get*, *set* and *toggle* commands all use
this API. Usage of this API is also the recommended approach when writing custom
tools (e.g.: switching the current mode based on the input from a light sensor).

## Third party integrations

For Emacs users, a third party package exists to integrate darkman with Emacs:

	https://github.com/grtcdr/darkman.el

There also exists a plugin for neovim users:

	https://github.com/4e554c4c/darkman.nvim

# LOCATION

The current location may be specified in the configuration file. The location is
used to calculate what time sundown and sunrise happen.

It is also possible for darkman to automatically determine the system's location
using *geoclue*. Geoclue's reliability varies depending on distribution and
desktop environment, as an agent often needs to be configured for it to work
properly.

If no location is known, automatic transitions are disabled.

# CONFIGURATION

A configuration file and all settings are optional. Configuration is read from
*~/.config/darkman/config.yaml* (or other paths defined in the XDG basedir
spec), and has the following format:

```
lat: 52.3
lng: 4.8
dbusserver: true
```

The following settings are available:

- *lat*, *lng*: Latitude and longitude respectively. This value will be used at
  start-up, but will later be superseded by whatever geoclue resolves (if
  enabled). More than one decimal point is generally not needed, as described
  in https://xkcd.com/2170/.

- *usegeoclue* (true/*false*): Whether to use a local geoclue instance to
  determine the current location. On some distributions/setups, this may
  require setting up a geoclue agent to function properly. Setting this to
  false without explicitly setting lat and lng disables automatic transitions
  entirely.

- *dbusserver* (*true*/false): Whether to expose the current mode via darkman's
  own D-Bus API. The command line tool uses this API to apply changes, so it
  will not work if this setting is disabled.

- *portal* (*true*/false): Whether to expose the current mode via the XDG settings
  portal D-Bus API. Many desktop application will read the current mode via the
  portal and respect what darkman is indicating.

# ENVIRONMENT

The following environment variables are also read and will override the
configuration file:

_DARKMAN_LAT_
	Overrides the latitude for the current location.

_DARKMAN_LNG_
	Overrides the longitude for the current location.

_DARKMAN_DBUSSERVER_
	Overrides whether to expose the current mode via D-Bus.

_XDG_CURRENT_DESKTOP_
	Darkman does not use this variable; it should be defined for the
	*xdg-desktop-portal* instead.

# PRIVACY

Darkman will trigger a darkmode/lightmode transition at sundown in the current
location. Any application that is running locally can record or transmit the
time of these transitions and attempt to extrapolate information related the
current location.

When a web browser applies this transition at the same time, open websites can
record this information too.

A potential stalker or tracker can use the above information to infer that you
are likely in a region of the world where sunset happened at a specific time.
This region is usually a wide area spanning tens of thousands of kilometers, but
can be smaller for certain geographical locations.

The author of this tool uses a manually configured location with an integer
latitude and longitude to achieve a sensible balance between privacy and
convenience.

# DEVELOPMENT

For issues and general development inquiries, see the project home currently
hosted at GitLab:

	https://gitlab.com/WhyNotHugo/darkman

# DEBUGGING

To confirm which value is relayed via the *xdg-desktop-portal* use:

```
gdbus call --session \\
    --dest org.freedesktop.portal.Desktop \\
    --object-path /org/freedesktop/portal/desktop \\
    --method org.freedesktop.portal.Settings.ReadOne \\
    org.freedesktop.appearance color-scheme
```

# SEE ALSO

portals.conf(5)
gammastep(1)

# AUTHORS

Developed by Hugo O. Barrera <hugo@whynothugo.nl>, with invaluable contributions
from the community.

darkman is an open source project licensed under the ISC licence and developed
for anyone to use freely. If you would like to sponsor this project, see:

	https://whynothugo.nl/sponsor


CHANGELOG.md
# darkman changelog

## v0.1.0

Initial release

## v0.1.1

Include a LICENCE file.

## v0.1.2

Include a systemd service file.

## v0.1.3

Make geoclue stop polling when not needed any more.

## v0.2.0

- Redesign and rewrite.
- Cache the last known location. This makes subsequent startups faster.

## v0.3.0

Rewrite in golang. This makes building and redistributing a lot simpler, but
also allows switching to libraries which simplify exposing a D-Bus service.

## v0.4.0

Provide a D-Bus service for other applications to determine the current mode.
There's both a property (to poll the current mode) and a signal. The signal will
emit when the current mode changes and allows applications to react immediately.

## v0.4.1

- Logging improvements.
- Don't trigger a transition (or the `ModeChanged` signal) when the mode doesn't
  need to change.
- Run all transition scripts in parallel.

## v0.5.0

- darkman is now configurable via a configuration file, or environment
  variables. See README.md for details.
- `boottimer` has been moved into a separate module. It is considered usable by
  third parties, though its design is likely non-final, and it still needs finer
  testing outside our specific use-case.

## v0.5.1

- A man page is not included.

## v0.5.2

- Add a warning if geoclue is not responding. This should help debug instances
  where darkman isn't working because it can't figure out the current location.
- boottimer: Improve precision of the timer.
- Fix negative latitudes and longitudes not working.

## v0.5.3

- Changing the current mode via the D-Bus API is now possible.
- A `darkmanctl` contrib script is now included to manually transition the
  current mode.

## v0.6.0

- A bug where the D-Bus server sometimes failed to start has been fixed.
- `darkmanctl` is now installed by default, along with shell completions for it.
- A go package is included to query and control the current mode from other go
  applications.

## v0.7.0

- `darkmanctl` has been merged into separate commands of `darkman`. This avoids
  installing a second binary, reduces installation size around half, and the new
  command is shorter to type. If you run darkman via an init script or alike,
  update it to execute `darkmanctl run` instead of just `darkman`.
- How we interact with geoclue has been simplified a bit -- we delegate more to
  geoclue rather than such continuos control over it.
- Calculation of next sundown and sunrise has been simplified, as well as
  reducing unnecessary calculations.

## 0.7.1

- Avoid scripts running concurrently (this could only happen in some very
  unusual wake-up scenarios with very specific timing).

## 1.0.0

- Support integration with the `xdg-desktop-portal`' [dark style preference
  setting][xdp]. With this integration, applications using this portal will also
  be aligned by darkman's current mode and transitions.
- Scripts are now run in sequence rather than in parallel.

[xdp]: https://github.com/flatpak/xdg-desktop-portal/issues/629

## 1.0.1

- Minor bugfixes.
- Example scripts have been moved into `examples/`.

## 1.1.0

- The systemd service now has some extra hardening security rules. If you find
  any regression with your own scripts, please open an issue.
- It is now possible to run darkman without a location and without automated
  transitions (this may be useful when controlling it via a light sensor, or
  manually). See the man page for details.

## 1.2.0

- Fix a signal not being raised when changing the value via the
  xdg-desktop-portal. This resulted in applications not immediately picking up
  the change.

## 1.3.0

- The start-up sequence has changed slightly. Previously, darkman did not listen
  on the D-Bus APIs until a mode had been set. This was problematic in scenarios
  where geoclue does not work and no location ever resolved; because no
  transition ever happened, darkman never listened on the D-Bus APIs, which
  implied that start-up had not been successful. Darkman will now bind to the
  D-Bus APIs immediately on start-up, but only emit a change when an actual
  transition happens. Because of this, querying the mode during start-up _may_
  return `NULL`, whereas previously it would simply not respond until another
  mode was set or until the query timed out.
- Disabling automatic transitions is now supported.
- Darkman will now cache the last mode to disk. If location-based transitions
  are disabled, this mode will be used at start-up.
- The documentation is now also available at https://darkman.whynothugo.nl/

## 1.3.1

- Fix how the signal for the XDG portal is raised. Previously it was raised
  directly, which would would only be picked up by specific unsandboxed
  applications, but not most of them (mostly since `darkman` raised the signal
  itself rather than via the desktop portal).

## 1.4.0

- The man page now indicates the default value for each config setting.
- When failing to register a D-Bus service (e.g.: because it is already taken),
  darkman will not exit immediately, rather than simply log the error.

## 1.5.0

- When running via systemd.service, darkman no longer depends on
  `graphical-session.target`.
- Implemented `--version`.
- Substantially trimmed down dependency tree by replacing `cobra` and `viper`
  with `flaggy`

## 1.5.1

- Fixed breakage in build scripts.

## 1.5.2

- Fixed build failures on non-64bit architectures.

## 1.5.3

- Avoid conflicting usage of LDFLAGS in build scripts.

## 1.5.4

- Fixed a bug where darkman would stall if geoclue was not running or not
  present.

## 2.0.0

- **BREAKING** Exit with an error if the configuration file has unknown fields.
- **BREAKING** Geoclue integration is now disabled by default. To retain the
  previous behaviour, explicitly enable it in the configuration file.
- **BREAKING** Go version 1.18 is now required to build.
- **BREAKING** The deprecated `darkmanctl` alias is gone. Use `darkman` instead.

darkman.py
from __future__ import annotations
import subprocess

import logging
from dataclasses import dataclass
from datetime import datetime
from datetime import timedelta
from enum import Enum
from pathlib import Path
from typing import Callable
from typing import Optional

from astral import Observer
from astral.sun import sun
from dateutil.tz import tzlocal
from twisted.internet import defer
from twisted.internet import reactor
from txdbus import client
from txdbus import error
from txdbus.objects import RemoteDBusObject
from xdg import BaseDirectory

logger = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)


class Mode(Enum):
    Light = "light-mode"
    Dark = "dark-mode"

    @property
    def opposite(mode: Mode) -> Mode:
        if mode == Mode.Light:
            return Mode.Dark
        elif mode == Mode.Dark:
            return Mode.Light
        return ValueError("Expected a Mode.")

    def activate(self):
        """Activate this mode.

        This is done by running one of more scripts in directories defined by
        convention:

        """
        logger.info("Activating %s.", self)

        scripts = {}
        for d in BaseDirectory.xdg_data_dirs:
            path = Path(d) / f"{self.value}.d"

            if path.is_dir():
                for file in (path).iterdir():
                    if file.name not in scripts:
                        scripts[file.name] = file
                        logger.info("Collected `%s.`", file)
                    else:
                        logger.info("Ignoring `%s`; it's been masked.", file)

        for name, script in scripts.items():
            proc = subprocess.run([script], shell=True)
            logger.info("Running %s returned %d.", name, proc.returncode)

        logger.info("Done switching to %s.", self)


@dataclass
class Event:
    """An event in the future in which we'll do a transition."""

    when: datetime
    mode: Mode

    def schedule(self, scheduler: Scheduler):
        """Schedules this event."""
        now = datetime.now(tzlocal())

        # max here set this to "0". This is to avoid breakage if we're scheduling during
        # the exact moment that a transition must take place.
        wait_for = max((self.when - now).total_seconds(), 0)

        logger.info("Will change to %s in %s seconds.", self.mode, wait_for)
        reactor.callLater(wait_for, self.execute, scheduler=scheduler)

    def execute(self, scheduler: GeoClueClient):
        """Execute this event, and schedule the next one."""
        self.mode.activate()
        scheduler.mode = self.mode

        # XXX: Should I wait here or have some offset? If the clock is skewed a few
        # seconds, this might re-schedule the same event....?
        next_event = self.gen_next(scheduler.location)
        next_event.schedule(scheduler=scheduler)

    @classmethod
    def gen_next(cls, location: Observer, date=None) -> Event:
        """Return the next event."""

        local_sun = sun(location, date=date, tzinfo=tzlocal())

        light_time = local_sun["dawn"]
        dark_time = local_sun["sunset"] + (local_sun["dusk"] - local_sun["sunset"])

        now = datetime.now(tzlocal())

        # XXX: There's an assumption made in this code that sunrise always comes before
        # sunset. I _think_ this is true anywhere in the world any time of the
        # year, though have a feeling that this might be one of these silly non-truths
        # we programmers assume somehow. 🤔

        if dark_time < now:
            # Already dark today, next change is tomorrow:
            return cls.gen_next(location, now + timedelta(days=1))
        elif light_time < now < dark_time:
            return Event(dark_time, Mode.Dark)
        elif now < light_time:
            return Event(light_time, Mode.Light)
        else:
            raise Exception("Something went wrong. Please report this!")


class Scheduler:
    _location: Optional[Observer] = None
    mode = Optional[Mode]

    @property
    def location(self):
        return self._location

    def set_location(self, location):
        if location != self._location:
            self._location = location
            self._set_timer()

    def _set_timer(self) -> None:
        """Set timers for the next color scheme transition."""

        for call in reactor.getDelayedCalls():
            # Cancel previously scheduled events.
            # We've moved, so those no longer apply.
            call.cancel()

        event = Event.gen_next(self.location)

        # Activate the opposite now. E.g.: If the next change is a transition to
        # dark mode, then we should be in light mode now.
        event.mode.opposite.activate()

        event.schedule(scheduler=self)

    # gsettings set io.elementary.terminal.settings prefer-dark-style true


class GeoClueClient:
    """A client for geoclue2 that waits for the location and runs the callback."""

    @defer.inlineCallbacks
    def _onLocationUpdated(self, old_path: str, new_path: str):
        """Work with location data to set timers.

        This function is called after GeoClue confirms our location, and sets timers to
        execute sunrise / sundown actions.
        """
        location_obj = yield self.connection.getRemoteObject(
            "org.freedesktop.GeoClue2",
            new_path,
        )
        lat = yield location_obj.callRemote(
            "Get",
            "org.freedesktop.GeoClue2.Location",
            "Latitude",
        )
        lon = yield location_obj.callRemote(
            "Get",
            "org.freedesktop.GeoClue2.Location",
            "Longitude",
        )
        logger.info("Got updated location data: %f, %f", lat, lon)
        location = Observer(latitude=lat, longitude=lon)

        self.callback(location)

    @defer.inlineCallbacks
    def _create_geoclue_client(self) -> RemoteDBusObject:
        """Creates a geoclue client, and returns it.

        Clients are private and per-connection. So we need to keep the connection around
        to further communicate with the client using it.
        """

        # Ask the manager API to create a client
        manager = yield self.connection.getRemoteObject(
            "org.freedesktop.GeoClue2",
            "/org/freedesktop/GeoClue2/Manager",
        )
        client_path = yield manager.callRemote("GetClient")

        # Get the client object:
        client = yield self.connection.getRemoteObject(
            "org.freedesktop.GeoClue2",
            client_path,
        )
        # This value needs to be set for some form of authorisation.
        # I've no idea what the _right_ value is, but this works fine.
        # Asked upstream at https://gitlab.freedesktop.org/geoclue/geoclue/-/issues/138
        yield client.callRemote(
            "Set",
            "org.freedesktop.GeoClue2.Client",
            "DesktopId",
            "9",
        )
        return client

    @defer.inlineCallbacks
    def main(self, callback: Callable):
        """Listens to location changes."""
        self.callback = callback

        try:
            # Connect to DBus and keep this connection.
            # We need to re-use the same connection for all calls.
            self.connection = yield client.connect(reactor, "system")

            # Get a geoclue client.
            geoclue_client = yield self._create_geoclue_client()
            logger.info("Got geoclue client: %s.", geoclue_client)

            # Set a callback for location updates.
            # This needs to be called _at least_ once.
            geoclue_client.notifyOnSignal("LocationUpdated", self._onLocationUpdated)

            yield geoclue_client.callRemote("Start")
            logger.info("Geoclue client started")

        except error.DBusException:
            logger.exception("DBus Error!")
        except Exception:
            logger.exception("Internal error!")


if __name__ == "__main__":
    scheduler = Scheduler()
    geoclient = GeoClueClient()
    reactor.callWhenRunning(geoclient.main, scheduler.set_location)
    reactor.run()
